File: tests/pairControl.test.ts
import { readFile } from "node:fs/promises";
import path from "node:path";
import { describe, expect, it } from "vitest";
import { PairControlRecordV1Schema } from "../src/shared/schemas/remoteAccess.js";
import {
  PAIR_CONTROL_RECORD_MAX_BYTES,
  PairControlIngressStateV1,
  encodePairControlSignatureInputV1,
  pairControlTransportAllows,
  parseAndVerifyPairControlRecordV1,
  serializePairControlPayloadV1,
  serializePairControlRecordV1,
  verifyPairRevocationAckMacV1,
  verifyPairRevocationMacV1,
  type PairControlTransportV1
} from "../src/shared/pairControl.js";
import {
  createPairControlFixtureSet,
  serializePairControlFixture
} from "../src/shared/pairControlContract.js";

type JsonObject = Record<string, unknown>;

function object(value: unknown, name = "fixture value"): JsonObject {
  if (value === null || typeof value !== "object" || Array.isArray(value)) {
    throw new TypeError(`${name} must be an object.`);
  }
  return value as JsonObject;
}

function array(value: unknown, name = "fixture value"): unknown[] {
  if (!Array.isArray(value)) {
    throw new TypeError(`${name} must be an array.`);
  }
  return value;
}

function string(value: unknown, name = "fixture value"): string {
  if (typeof value !== "string") {
    throw new TypeError(`${name} must be a string.`);
  }
  return value;
}

function number(value: unknown, name = "fixture value"): number {
  if (typeof value !== "number") {
    throw new TypeError(`${name} must be a number.`);
  }
  return value;
}

function fixture(): JsonObject {
  return object(createPairControlFixtureSet().values().next().value, "pair-control fixture");
}

function recordVectors(value: JsonObject): JsonObject[] {
  return array(value.records, "record vectors").map((record) => object(record, "record vector"));
}

function verificationOptions(value: JsonObject, record: JsonObject) {
  const publicKeys = object(value.installationPublicKeys, "installation public keys");
  const parsed = PairControlRecordV1Schema.parse(record.value);
  return {
    installationPublicKey: Buffer.from(
      string(parsed.side === 1 ? publicKeys.host : publicKeys.remote),
      "base64url"
    ),
    expectedPairId: string(object(value.context).pairId) as typeof parsed.pairId,
    expectedSide: parsed.side,
    nowSeconds: BigInt(string(value.acceptedAt)),
    timestampMode: "worker_ingress" as const,
    transport: string(record.ingressTransport) as PairControlTransportV1
  };
}

describe("PairControlRecordV1", () => {
  it("recreates the public fixture byte-for-byte", async () => {
    const fixtures = createPairControlFixtureSet();
    const expected = fixtures.get("fixtures/crypto/pair-control-record-v1.json");
    expect(expected).toBeDefined();
    const actual = await readFile(path.join(
      process.cwd(),
      "contracts",
      "remote",
      "v1",
      "fixtures",
      "crypto",
      "pair-control-record-v1.json"
    ), "utf8");
    expect(actual).toBe(serializePairControlFixture(expected as never));
  });

  it("accepts all nine exact signed records and pins payload/signature bytes", () => {
    const value = fixture();
    const records = recordVectors(value);
    expect(records.map((record) => number(record.typeByte))).toEqual([1, 2, 3, 4, 5, 6, 7, 8, 9]);
    for (const vector of records) {
      const payload = Buffer.from(string(vector.canonicalBytesB64), "base64url");
      const parsed = parseAndVerifyPairControlRecordV1(
        payload,
        verificationOptions(value, vector)
      );
      expect(parsed).toEqual(PairControlRecordV1Schema.parse(vector.value));
      expect(serializePairControlRecordV1(parsed)).toEqual(payload);
      expect(serializePairControlPayloadV1(parsed)).toEqual(
        Buffer.from(string(vector.payloadBytesB64), "base64url")
      );
      expect(encodePairControlSignatureInputV1(parsed)).toEqual(
        Buffer.from(string(vector.signatureInputB64), "base64url")
      );
    }
  });

  it("rejects structural, canonical, signature, context, hash, and size substitutions", () => {
    const value = fixture();
    const records = recordVectors(value);
    const baseOptions = verificationOptions(value, records[0]);
    for (const raw of array(value.rejections, "rejections")) {
      const vector = object(raw, "rejection");
      const keySide = number(vector.keySide ?? vector.side ?? 1) as 1 | 2;
      const expectedSide = number(vector.expectedSide ?? vector.side ?? 1) as 1 | 2;
      const options = {
        ...baseOptions,
        installationPublicKey: Buffer.from(
          string(object(value.installationPublicKeys)[keySide === 1 ? "host" : "remote"]),
          "base64url"
        ),
        expectedSide
      };
      expect(
        () => parseAndVerifyPairControlRecordV1(
          Buffer.from(string(vector.payloadB64), "base64url"),
          options
        ),
        string(vector.name)
      ).toThrow(string(vector.errorCode));
    }
    const boundary = object(value.boundary, "boundary");
    const maximum = Buffer.from(string(boundary.maximumRecordB64), "base64url");
    expect(maximum.byteLength).toBe(number(boundary.maximumRecordBytes));
    expect(maximum.byteLength).toBeLessThanOrEqual(PAIR_CONTROL_RECORD_MAX_BYTES);
    expect(() => parseAndVerifyPairControlRecordV1(
      maximum,
      {
        ...baseOptions,
        installationPublicKey: Buffer.from(
          string(object(value.installationPublicKeys).host),
          "base64url"
        ),
        expectedSide: 1,
        transport: "websocket"
      }
    )).not.toThrow();
    expect(() => parseAndVerifyPairControlRecordV1(
      Buffer.from(string(boundary.overLimitPayloadB64), "base64url"),
      baseOptions
    )).toThrow("payload_too_large");
  });

  it("enforces the closed WebSocket, publish, revoke, ack, and poll matrix", () => {
    const expected: Record<PairControlTransportV1, number[]> = {
      websocket: [1, 2, 3, 4, 5, 6, 7, 8, 9],
      https_publish: [1, 2, 3, 4, 5, 6, 9],
      https_revoke: [7],
      https_revocation_ack: [8],
      https_poll: [1, 2, 3, 4, 5, 6, 7, 8, 9]
    };
    for (const [transport, allowedTypes] of Object.entries(expected) as Array<[
      PairControlTransportV1,
      number[]
    ]>) {
      for (let type = 1; type <= 9; type += 1) {
        expect(pairControlTransportAllows(transport, type as never)).toBe(allowedTypes.includes(type));
      }
    }
    expect(fixture().transportMatrix).toEqual(expected);
  });

  it("shares per-side high-water and replay state across transports and restart", () => {
    const value = fixture();
    const records = recordVectors(value);
    const publicKeys = object(value.installationPublicKeys);
    const stateOptions = {
      expectedPairId: string(object(value.context).pairId) as never,
      hostInstallationPublicKey: Buffer.from(string(publicKeys.host), "base64url"),
      remoteInstallationPublicKey: Buffer.from(string(publicKeys.remote), "base64url")
    };
    let state = new PairControlIngressStateV1(stateOptions);
    const nowSeconds = BigInt(string(value.acceptedAt));

    for (const vector of records.slice(0, 6)) {
      expect(state.accept(
        Buffer.from(string(vector.canonicalBytesB64), "base64url"),
        string(vector.ingressTransport) as PairControlTransportV1,
        nowSeconds
      )).toBe("accepted");
    }
    const revocation = records[6];
    expect(() => state.accept(
      Buffer.from(string(revocation.canonicalBytesB64), "base64url"),
      "https_publish",
      nowSeconds
    )).toThrow("wrong_transport");
    expect(state.accept(
      Buffer.from(string(revocation.canonicalBytesB64), "base64url"),
      "https_revoke",
      nowSeconds
    )).toBe("accepted");
    const acknowledgement = records[7];
    expect(state.accept(
      Buffer.from(string(acknowledgement.canonicalBytesB64), "base64url"),
      "https_revocation_ack",
      nowSeconds
    )).toBe("accepted");

    state = new PairControlIngressStateV1(stateOptions, state.snapshot());
    const errorRecord = records[8];
    const errorBytes = Buffer.from(string(errorRecord.canonicalBytesB64), "base64url");
    expect(state.accept(errorBytes, "https_publish", nowSeconds)).toBe("accepted");
    expect(state.accept(errorBytes, "websocket", nowSeconds)).toBe("idempotent");

    for (const raw of array(value.stateRejections, "state rejections")) {
      const vector = object(raw, "state rejection");
      expect(() => state.accept(
        Buffer.from(string(vector.payloadB64), "base64url"),
        string(vector.transport) as PairControlTransportV1,
        nowSeconds
      ), string(vector.name)).toThrow(string(vector.errorCode));
    }

    const transition = object(value.generationTransition, "generation transition");
    const advanceBytes = Buffer.from(string(transition.advancePayloadB64), "base64url");
    expect(state.accept(advanceBytes, "websocket", nowSeconds)).toBe("accepted");
    expect(state.accept(advanceBytes, "https_publish", nowSeconds)).toBe("idempotent");
    expect(() => state.accept(
      Buffer.from(string(transition.stalePayloadB64), "base64url"),
      "websocket",
      nowSeconds
    )).toThrow("stale_generation");
  });

  it("distinguishes first-ingress freshness from delayed durable delivery", () => {
    const value = fixture();
    const records = recordVectors(value);
    const capabilities = records[1];
    const capabilitiesBytes = Buffer.from(string(capabilities.canonicalBytesB64), "base64url");
    const options = verificationOptions(value, capabilities);
    const delayedAt = BigInt(string(value.delayedAt));
    expect(() => parseAndVerifyPairControlRecordV1(capabilitiesBytes, {
      ...options,
      nowSeconds: delayedAt
    })).toThrow("timestamp_out_of_window");
    expect(() => parseAndVerifyPairControlRecordV1(capabilitiesBytes, {
      ...options,
      nowSeconds: delayedAt,
      timestampMode: "durable_delivery",
      transport: "https_poll"
    })).not.toThrow();

    const presence = records[4];
    expect(() => parseAndVerifyPairControlRecordV1(
      Buffer.from(string(presence.canonicalBytesB64), "base64url"),
      {
        ...verificationOptions(value, presence),
        nowSeconds: delayedAt,
        timestampMode: "durable_delivery",
        transport: "https_poll"
      }
    )).toThrow("presence_expired");
  });

  it("verifies domain-separated revocation and acknowledgement MACs only at helpers", () => {
    const value = fixture();
    const revocation = object(value.revocation, "revocation fixture");
    const context = object(revocation.context, "revocation context");
    const key = Buffer.from(string(revocation.revocationKeyB64), "base64url");
    const records = recordVectors(value);
    const revokeRecord = PairControlRecordV1Schema.parse(records[6].value);
    const acknowledgement = PairControlRecordV1Schema.parse(records[7].value);
    expect(verifyPairRevocationMacV1(key, revokeRecord, context as never)).toBe(true);
    expect(verifyPairRevocationAckMacV1(key, acknowledgement, context as never)).toBe(true);

    for (const raw of array(revocation.rejections, "revocation rejections")) {
      const vector = object(raw, "revocation rejection");
      const record = PairControlRecordV1Schema.parse(vector.value);
      const vectorKey = Buffer.from(
        string(vector.keyB64 ?? revocation.revocationKeyB64),
        "base64url"
      );
      const vectorContext = object(vector.context ?? context, "revocation vector context");
      const verified = record.type === 7
        ? verifyPairRevocationMacV1(vectorKey, record, vectorContext as never)
        : verifyPairRevocationAckMacV1(vectorKey, record, vectorContext as never);
      expect(verified, string(vector.name)).toBe(false);
    }

    const wrongMac = object(revocation.workerOpaqueWrongMac, "opaque wrong MAC");
    const wrongMacRecord = PairControlRecordV1Schema.parse(wrongMac.value);
    const wrongMacBytes = Buffer.from(string(wrongMac.canonicalBytesB64), "base64url");
    expect(() => parseAndVerifyPairControlRecordV1(
      wrongMacBytes,
      verificationOptions(value, {
        ...records[6],
        value: wrongMacRecord
      })
    )).not.toThrow();
    expect(() => parseAndVerifyPairControlRecordV1(wrongMacBytes, {
      ...verificationOptions(value, {
        ...records[6],
        value: wrongMacRecord
      }),
      revocation: { key, context: context as never }
    })).toThrow("invalid_revocation_mac");
  });
});

File: contracts/remote/v1/conformance-go/conformance/pair_control_test.go
package conformance_test

import (
	"bytes"
	"encoding/json"
	"errors"
	"os"
	"path/filepath"
	"reflect"
	"strconv"
	"testing"

	"github.com/waifucave/discord-waifus/contracts/remote/v1/conformance-go/internal/pairing"
	"github.com/waifucave/discord-waifus/contracts/remote/v1/conformance-go/internal/vectors"
)

type controlRecordVector struct {
	Name              string                       `json:"name"`
	TypeByte          int                          `json:"typeByte"`
	IngressTransport  pairing.PairControlTransport `json:"ingressTransport"`
	Value             pairing.PairControlRecord    `json:"value"`
	CanonicalBytesB64 string                       `json:"canonicalBytesB64"`
	PayloadBytesB64   string                       `json:"payloadBytesB64"`
	SignatureInputB64 string                       `json:"signatureInputB64"`
}

type controlRejectionVector struct {
	Name         string                       `json:"name"`
	PayloadB64   string                       `json:"payloadB64"`
	ErrorCode    string                       `json:"errorCode"`
	Side         int                          `json:"side"`
	KeySide      int                          `json:"keySide"`
	ExpectedSide int                          `json:"expectedSide"`
	Transport    pairing.PairControlTransport `json:"transport"`
}

type revocationRejectionVector struct {
	Name    string                         `json:"name"`
	Value   pairing.PairControlRecord      `json:"value"`
	KeyB64  string                         `json:"keyB64"`
	Context *pairing.PairRevocationContext `json:"context"`
}

type pairControlFixture struct {
	AcceptedAt             string `json:"acceptedAt"`
	DelayedAt              string `json:"delayedAt"`
	InstallationPublicKeys struct {
		Host   string `json:"host"`
		Remote string `json:"remote"`
	} `json:"installationPublicKeys"`
	Context struct {
		PairID string `json:"pairId"`
	} `json:"context"`
	Records              []controlRecordVector    `json:"records"`
	Rejections           []controlRejectionVector `json:"rejections"`
	StateRejections      []controlRejectionVector `json:"stateRejections"`
	GenerationTransition struct {
		AdvancePayloadB64 string `json:"advancePayloadB64"`
		StalePayloadB64   string `json:"stalePayloadB64"`
	} `json:"generationTransition"`
	Boundary struct {
		MaximumRecordBytes  int    `json:"maximumRecordBytes"`
		MaximumRecordB64    string `json:"maximumRecordB64"`
		OverLimitPayloadB64 string `json:"overLimitPayloadB64"`
	} `json:"boundary"`
	Revocation struct {
		RevocationKeyB64         string                        `json:"revocationKeyB64"`
		Context                  pairing.PairRevocationContext `json:"context"`
		RevocationMACInputB64    string                        `json:"revocationMacInputB64"`
		RevocationAckMACInputB64 string                        `json:"revocationAckMacInputB64"`
		Rejections               []revocationRejectionVector   `json:"rejections"`
		WorkerOpaqueWrongMAC     controlRecordVector           `json:"workerOpaqueWrongMac"`
	} `json:"revocation"`
}

func pairControlFixtureBytes(t *testing.T) []byte {
	t.Helper()
	value, err := os.ReadFile(filepath.Join("..", "..", "fixtures", "crypto", "pair-control-record-v1.json"))
	if err != nil {
		t.Fatalf("read committed pair control fixture: %v", err)
	}
	return value
}

func decodePairControlFixture(t *testing.T) pairControlFixture {
	t.Helper()
	decoder := json.NewDecoder(bytes.NewReader(pairControlFixtureBytes(t)))
	decoder.UseNumber()
	var fixture pairControlFixture
	if err := decoder.Decode(&fixture); err != nil {
		t.Fatalf("decode pair control fixture: %v", err)
	}
	return fixture
}

func controlErrorCode(t *testing.T, err error) string {
	t.Helper()
	var protocolError *pairing.ControlError
	if !errors.As(err, &protocolError) {
		t.Fatalf("expected ControlError, got %T: %v", err, err)
	}
	return protocolError.Code
}

func fixtureSeconds(t *testing.T, value string) uint64 {
	t.Helper()
	parsed, err := strconv.ParseUint(value, 10, 64)
	if err != nil {
		t.Fatalf("parse fixture timestamp: %v", err)
	}
	return parsed
}

func controlPublicKey(t *testing.T, fixture pairControlFixture, side int) []byte {
	t.Helper()
	if side == 1 {
		return decodeBase64URL(t, fixture.InstallationPublicKeys.Host)
	}
	return decodeBase64URL(t, fixture.InstallationPublicKeys.Remote)
}

func controlVerifyOptions(t *testing.T, fixture pairControlFixture, vector controlRecordVector) pairing.PairControlVerifyOptions {
	t.Helper()
	return pairing.PairControlVerifyOptions{
		InstallationPublicKey: controlPublicKey(t, fixture, vector.Value.Side),
		ExpectedPairID:        fixture.Context.PairID,
		ExpectedSide:          vector.Value.Side,
		NowSeconds:            fixtureSeconds(t, fixture.AcceptedAt),
		TimestampMode:         pairing.ControlWorkerIngress,
		Transport:             vector.IngressTransport,
		ExpectedProtocolMajor: 1,
		ExpectedProtocolMinor: 0,
	}
}

func TestGoGeneratorMatchesCommittedPairControlFixture(t *testing.T) {
	expected, err := vectors.BuildPairControlV1JSON()
	if err != nil {
		t.Fatalf("build pair control fixture: %v", err)
	}
	if !bytes.Equal(expected, pairControlFixtureBytes(t)) {
		t.Fatalf("Go-generated pair control fixture differs from committed bytes")
	}
}

func TestPairControlValidRecordsAndSignatureInputs(t *testing.T) {
	fixture := decodePairControlFixture(t)
	if len(fixture.Records) != 9 {
		t.Fatalf("record count = %d, want 9", len(fixture.Records))
	}
	for index, vector := range fixture.Records {
		if vector.TypeByte != index+1 {
			t.Fatalf("record %d type = %d", index, vector.TypeByte)
		}
		t.Run(vector.Name, func(t *testing.T) {
			payload := decodeBase64URL(t, vector.CanonicalBytesB64)
			parsed, err := pairing.ParseAndVerifyPairControlRecord(payload, controlVerifyOptions(t, fixture, vector))
			if err != nil {
				t.Fatalf("parse valid control record: %v", err)
			}
			if !reflect.DeepEqual(parsed, vector.Value) {
				t.Fatalf("parsed record differs from fixture")
			}
			canonical, err := pairing.CanonicalPairControlJSON(parsed)
			if err != nil || !bytes.Equal(canonical, payload) {
				t.Fatalf("canonical bytes differ: %v", err)
			}
			payloadBytes, err := pairing.PairControlPayloadJSON(parsed)
			if err != nil || !bytes.Equal(payloadBytes, decodeBase64URL(t, vector.PayloadBytesB64)) {
				t.Fatalf("payload bytes differ: %v", err)
			}
			signatureInput, err := pairing.PairControlSignatureInput(parsed)
			if err != nil || !bytes.Equal(signatureInput, decodeBase64URL(t, vector.SignatureInputB64)) {
				t.Fatalf("signature input differs: %v", err)
			}
		})
	}
}

func TestPairControlRejectionsAndBoundary(t *testing.T) {
	fixture := decodePairControlFixture(t)
	for _, vector := range fixture.Rejections {
		vector := vector
		t.Run(vector.Name, func(t *testing.T) {
			keySide := vector.KeySide
			if keySide == 0 {
				keySide = vector.Side
			}
			if keySide == 0 {
				keySide = 1
			}
			expectedSide := vector.ExpectedSide
			if expectedSide == 0 {
				expectedSide = vector.Side
			}
			if expectedSide == 0 {
				expectedSide = 1
			}
			transport := vector.Transport
			if transport == "" {
				transport = pairing.ControlHTTPSPublish
			}
			_, err := pairing.ParseAndVerifyPairControlRecord(
				decodeBase64URL(t, vector.PayloadB64),
				pairing.PairControlVerifyOptions{
					InstallationPublicKey: controlPublicKey(t, fixture, keySide),
					ExpectedPairID:        fixture.Context.PairID, ExpectedSide: expectedSide,
					NowSeconds: fixtureSeconds(t, fixture.AcceptedAt), TimestampMode: pairing.ControlWorkerIngress,
					Transport: transport, ExpectedProtocolMajor: 1, ExpectedProtocolMinor: 0,
				},
			)
			if err == nil || controlErrorCode(t, err) != vector.ErrorCode {
				t.Fatalf("error = %v, want %s", err, vector.ErrorCode)
			}
		})
	}
	maximum := decodeBase64URL(t, fixture.Boundary.MaximumRecordB64)
	if len(maximum) != fixture.Boundary.MaximumRecordBytes || len(maximum) > pairing.PairControlRecordMax {
		t.Fatalf("maximum record boundary differs: %d", len(maximum))
	}
	_, err := pairing.ParseAndVerifyPairControlRecord(maximum, pairing.PairControlVerifyOptions{
		InstallationPublicKey: controlPublicKey(t, fixture, 1), ExpectedPairID: fixture.Context.PairID, ExpectedSide: 1,
		NowSeconds: fixtureSeconds(t, fixture.AcceptedAt), TimestampMode: pairing.ControlWorkerIngress,
		Transport: pairing.ControlWebSocket, ExpectedProtocolMajor: 1, ExpectedProtocolMinor: 0,
	})
	if err != nil {
		t.Fatalf("maximum record rejected: %v", err)
	}
	_, err = pairing.ParseAndVerifyPairControlRecord(
		decodeBase64URL(t, fixture.Boundary.OverLimitPayloadB64),
		controlVerifyOptions(t, fixture, fixture.Records[0]),
	)
	if err == nil || controlErrorCode(t, err) != "payload_too_large" {
		t.Fatalf("over-limit record outcome = %v", err)
	}
}

func TestPairControlTransportMatrix(t *testing.T) {
	expected := map[pairing.PairControlTransport][]int{
		pairing.ControlWebSocket:      {1, 2, 3, 4, 5, 6, 7, 8, 9},
		pairing.ControlHTTPSPublish:   {1, 2, 3, 4, 5, 6, 9},
		pairing.ControlHTTPSRevoke:    {7},
		pairing.ControlHTTPSRevokeAck: {8},
		pairing.ControlHTTPSPoll:      {1, 2, 3, 4, 5, 6, 7, 8, 9},
	}
	for transport, allowed := range expected {
		for typeValue := 1; typeValue <= 9; typeValue++ {
			want := false
			for _, candidate := range allowed {
				want = want || candidate == typeValue
			}
			if pairing.PairControlTransportAllows(transport, typeValue) != want {
				t.Fatalf("transport %s type %d differs", transport, typeValue)
			}
		}
	}
}

func TestPairControlSharedHighWaterRestartAndGeneration(t *testing.T) {
	fixture := decodePairControlFixture(t)
	state, err := pairing.NewPairControlIngress(
		fixture.Context.PairID,
		controlPublicKey(t, fixture, 1),
		controlPublicKey(t, fixture, 2),
		nil,
	)
	if err != nil {
		t.Fatalf("create ingress state: %v", err)
	}
	now := fixtureSeconds(t, fixture.AcceptedAt)
	for _, vector := range fixture.Records[:6] {
		outcome, err := state.Accept(decodeBase64URL(t, vector.CanonicalBytesB64), vector.IngressTransport, now)
		if err != nil || outcome != "accepted" {
			t.Fatalf("accept %s: outcome=%s err=%v", vector.Name, outcome, err)
		}
	}
	revocation := fixture.Records[6]
	if _, err := state.Accept(decodeBase64URL(t, revocation.CanonicalBytesB64), pairing.ControlHTTPSPublish, now); err == nil || controlErrorCode(t, err) != "wrong_transport" {
		t.Fatalf("wrong-route revocation outcome = %v", err)
	}
	if outcome, err := state.Accept(decodeBase64URL(t, revocation.CanonicalBytesB64), pairing.ControlHTTPSRevoke, now); err != nil || outcome != "accepted" {
		t.Fatalf("accept revocation after wrong route: outcome=%s err=%v", outcome, err)
	}
	ack := fixture.Records[7]
	if outcome, err := state.Accept(decodeBase64URL(t, ack.CanonicalBytesB64), pairing.ControlHTTPSRevokeAck, now); err != nil || outcome != "accepted" {
		t.Fatalf("accept revocation ack: outcome=%s err=%v", outcome, err)
	}
	snapshot := state.Snapshot()
	state, err = pairing.NewPairControlIngress(
		fixture.Context.PairID,
		controlPublicKey(t, fixture, 1),
		controlPublicKey(t, fixture, 2),
		&snapshot,
	)
	if err != nil {
		t.Fatalf("restore ingress state: %v", err)
	}
	errorVector := fixture.Records[8]
	errorBytes := decodeBase64URL(t, errorVector.CanonicalBytesB64)
	if outcome, err := state.Accept(errorBytes, pairing.ControlHTTPSPublish, now); err != nil || outcome != "accepted" {
		t.Fatalf("accept post-restart error: outcome=%s err=%v", outcome, err)
	}
	if outcome, err := state.Accept(errorBytes, pairing.ControlWebSocket, now); err != nil || outcome != "idempotent" {
		t.Fatalf("cross-transport retry: outcome=%s err=%v", outcome, err)
	}
	for _, vector := range fixture.StateRejections {
		if _, err := state.Accept(decodeBase64URL(t, vector.PayloadB64), vector.Transport, now); err == nil || controlErrorCode(t, err) != vector.ErrorCode {
			t.Fatalf("state rejection %s = %v", vector.Name, err)
		}
	}
	advance := decodeBase64URL(t, fixture.GenerationTransition.AdvancePayloadB64)
	if outcome, err := state.Accept(advance, pairing.ControlWebSocket, now); err != nil || outcome != "accepted" {
		t.Fatalf("advance generation: outcome=%s err=%v", outcome, err)
	}
	if outcome, err := state.Accept(advance, pairing.ControlHTTPSPublish, now); err != nil || outcome != "idempotent" {
		t.Fatalf("retry generation advance: outcome=%s err=%v", outcome, err)
	}
	if _, err := state.Accept(decodeBase64URL(t, fixture.GenerationTransition.StalePayloadB64), pairing.ControlWebSocket, now); err == nil || controlErrorCode(t, err) != "stale_generation" {
		t.Fatalf("stale generation outcome = %v", err)
	}
}

func TestPairControlDelayedDeliveryAndRevocationMACs(t *testing.T) {
	fixture := decodePairControlFixture(t)
	delayedAt := fixtureSeconds(t, fixture.DelayedAt)
	capabilities := fixture.Records[1]
	capabilitiesBytes := decodeBase64URL(t, capabilities.CanonicalBytesB64)
	options := controlVerifyOptions(t, fixture, capabilities)
	options.NowSeconds = delayedAt
	if _, err := pairing.ParseAndVerifyPairControlRecord(capabilitiesBytes, options); err == nil || controlErrorCode(t, err) != "timestamp_out_of_window" {
		t.Fatalf("old first-ingress record outcome = %v", err)
	}
	options.TimestampMode = pairing.ControlDurableDelivery
	options.Transport = pairing.ControlHTTPSPoll
	if _, err := pairing.ParseAndVerifyPairControlRecord(capabilitiesBytes, options); err != nil {
		t.Fatalf("delayed durable record rejected: %v", err)
	}
	presence := fixture.Records[4]
	presenceOptions := controlVerifyOptions(t, fixture, presence)
	presenceOptions.NowSeconds = delayedAt
	presenceOptions.TimestampMode = pairing.ControlDurableDelivery
	presenceOptions.Transport = pairing.ControlHTTPSPoll
	if _, err := pairing.ParseAndVerifyPairControlRecord(decodeBase64URL(t, presence.CanonicalBytesB64), presenceOptions); err == nil || controlErrorCode(t, err) != "presence_expired" {
		t.Fatalf("expired presence outcome = %v", err)
	}

	key := decodeBase64URL(t, fixture.Revocation.RevocationKeyB64)
	revocation := fixture.Records[6].Value
	ack := fixture.Records[7].Value
	if !pairing.VerifyPairRevocationMAC(key, revocation, fixture.Revocation.Context) || !pairing.VerifyPairRevocationMAC(key, ack, fixture.Revocation.Context) {
		t.Fatalf("valid revocation MACs were rejected")
	}
	revocationInput, err := pairing.PairRevocationMACInput(revocation, fixture.Revocation.Context)
	if err != nil || !bytes.Equal(revocationInput, decodeBase64URL(t, fixture.Revocation.RevocationMACInputB64)) {
		t.Fatalf("revocation MAC input differs: %v", err)
	}
	ackInput, err := pairing.PairRevocationAckMACInput(ack, fixture.Revocation.Context)
	if err != nil || !bytes.Equal(ackInput, decodeBase64URL(t, fixture.Revocation.RevocationAckMACInputB64)) {
		t.Fatalf("revocation ack MAC input differs: %v", err)
	}
	for _, vector := range fixture.Revocation.Rejections {
		vectorKey := key
		if vector.KeyB64 != "" {
			vectorKey = decodeBase64URL(t, vector.KeyB64)
		}
		context := fixture.Revocation.Context
		if vector.Context != nil {
			context = *vector.Context
		}
		if pairing.VerifyPairRevocationMAC(vectorKey, vector.Value, context) {
			t.Fatalf("revocation rejection %s was accepted", vector.Name)
		}
	}
	wrongMAC := fixture.Revocation.WorkerOpaqueWrongMAC
	workerOptions := controlVerifyOptions(t, fixture, wrongMAC)
	wrongMACBytes := decodeBase64URL(t, wrongMAC.CanonicalBytesB64)
	if _, err := pairing.ParseAndVerifyPairControlRecord(wrongMACBytes, workerOptions); err != nil {
		t.Fatalf("Worker-style opaque MAC validation rejected record: %v", err)
	}
	workerOptions.RevocationKey = key
	workerOptions.RevocationContext = &fixture.Revocation.Context
	if _, err := pairing.ParseAndVerifyPairControlRecord(wrongMACBytes, workerOptions); err == nil || controlErrorCode(t, err) != "invalid_revocation_mac" {
		t.Fatalf("helper-style wrong MAC outcome = %v", err)
	}
}
